### PR TITLE
feat: support customized DNS resolving for remote registry

### LINKED ISSUE #688: Support customized DNS resolving from a `host` to a specific `address`
### What is the version of your ORAS CLI

_No response_

### What would you like to be added?

I would like to have a new flag added to all remote registry related commands:
```
--resolve <host:port:address>
```

So any outgoing request towards `host`:`port` should be resolve to `address`:`port`.

### Why is this needed for ORAS?

So ORAS can be 1) used to debug registry services running locally; 2) hardcode DNS resolving to reduce DNS bouncing in geo-replicated registry services.

### Are you willing to submit PRs to contribute to this feature?

- [X] Yes, I am willing to implement it.

### PR DESCRIPTION
**Pull request by qweeah**
Resolves #688

### CODE REVIEW

File: cmd/oras/internal/option/remote.go [hunk 1]
```diff
@@ -109,21 +151,42 @@ func (opts *Remote) tlsConfig() (*tls.Config, error) {
...
+		if addr == fmt.Sprintf("%s:%d", r.from, r.port) {
+			matched = r
+			break
+		}
+	}
+	if matched == nil {
+		return defaultDialer.DialContext(ctx, network, addr)
+	}
+	return net.DialTCP(network, nil, &net.TCPAddr{IP: matched.to, Port: matched.port})
+}
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/remote.go:173]: The `net.DialTCP` does not accept a context. How about
```go
type resolveDialer struct {
	base   *net.Dialer
	resolve map[string]string
}

func (rd *resolveDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
	if resolved, ok := rd.resolve[addr]; ok {
		addr = resolved
	}
	return rd.base.DialContext(ctx, network, addr)
}
```
Will above code work?

Comment (qweeah) [reply cmd/oras/internal/option/remote.go:173]: It surely works, but using mapping surely reduces extensibility for supporting ipv6. Resolved accordingly.

File: cmd/oras/internal/option/remote.go [hunk 2]
```diff
@@ -35,15 +36,24 @@ import (
 	"oras.land/oras/internal/version"
 )
 
+type ResolveEntry struct {
```
[the remark this answers is not in the document]
Comment (qweeah) [reply cmd/oras/internal/option/remote.go]: removed and changed to map

File: cmd/oras/internal/option/remote.go (hunk 1 above)
Comment (qweeah) [reply cmd/oras/internal/option/remote.go]: Checked it's non-relavant.

File: cmd/oras/internal/option/remote.go [hunk 3]
```diff
@@ -109,21 +143,37 @@ func (opts *Remote) tlsConfig() (*tls.Config, error) {
...
+// DialContext connects to the addr on the named network using
+// the provided context.
+func (opts *Remote) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	for k := range opts.resolve {
+		if k == addr {
+			addr = opts.resolve[k]
+		}
+	}
+	return defaultDialer.DialContext(ctx, network, addr)
+}
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/remote.go:160]: Can we decouple the dialer with the opts?

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**